File: src/assets/nodes_builtin/intent_nano.js
// Intent·Nano 节点插件（type intent_nano）：NanoJev 判别式意图路由（内置 Intent 加强版）。
// intents 编辑器 = name + description 两列（description 直供模型判别）；temperature/threshold 数字控件。
// 出口 branch_N/default 与内置 intent 同构（workflow_editor 的端口生成条件含 intent_nano）。
EdFW.register({
  type: "intent_nano", label: "Intent·Nano",    icon: "🧭", category: "llm",
  section: "意图选项（name + 语义描述；判别式路由 branch_N）",
  defaults: {nodeMeta: {title: "意图路由·Nano"},
             inputs: {inputParameters: [
                        {name: "query", input: {type: "string", value: {type: "literal", content: ""}}},
                        {name: "temperature", input: {type: "number", value: {type: "literal", content: "1.0"}}},
                        {name: "threshold", input: {type: "number", value: {type: "literal", content: "0.35"}}}],
                      intents: [{name: "code", description: "要求编写、修改、调试代码"},
                                {name: "search", description: "要求检索资料、查文档、搜索信息"},
                                {name: "chat", description: "闲聊、打招呼或常识问答"}]},
             outputs: [{name: "intent", type: "string", description: "top1 意图名（低于阈值为空）", fixed: true},
                       {name: "confidence", type: "number", description: "top1 置信度", fixed: true},
                       {name: "probabilities", type: "object", description: "完整概率分布", fixed: true}]},
  params: [
    { key: "query", label: "query", widget: "input", ph: "待分类文本（ref 上游字段）",
      get(n) { return ipGet(n, "query"); }, set(n, v) { ipSet(n, "query", v); } },
    { key: "temperature", label: "温度", widget: "input", ph: "1.0",
      get(n) { return ipGet(n, "temperature"); }, set(n, v) { ipSet(n, "temperature", v); } },
    { key: "threshold", label: "阈值", widget: "input", ph: "0.35（top1 低于它走 default）",
      get(n) { return ipGet(n, "threshold"); }, set(n, v) { ipSet(n, "threshold", v); } },
    { key: "intents", label: "意图", widget: "custom",
      get(n) { return (n.data.inputs?.intents || []).map(x => x.name).join(","); },
      set() {},
      html(n) {
        const intents = n.data.inputs?.intents || [];
        let h = '';
        intents.forEach((it, i) => {
          h += `<div style="display:flex;gap:3px;margin:2px 0;align-items:center"><span style="font-size:10px;color:#8a9099">${i + 1}.</span>` +
               `<input value="${ext(it.name || '')}" placeholder="意图名" onchange="NODEP_INANO_set(${i},'name',this.value)" style="width:80px;font-size:11px">` +
               `<input value="${ext(it.description || '')}" placeholder="语义描述（供 NanoJev 判别）" onchange="NODEP_INANO_set(${i},'description',this.value)" style="flex:1;font-size:11px">` +
               `<button class="del" onclick="NODEP_INANO_del(${i})">×</button></div>`;
        });
        h += `<button onclick="NODEP_INANO_add()">+ 意图</button>`;
        h += `<div style="font-size:10px;color:#8a9099;margin-top:4px">NanoJev 判别式（本地 ~1s 零 token）：输出概率分布，top1 低于阈值走 default。出口与内置 Intent 同构。</div>`;
        return h;
      } },
  ],
  nodeH(n) { return Math.max(3, (n.data.inputs?.intents || []).length) * 14 + 20; },
  body(n, g) {
    const intents = n.data.inputs?.intents || [];
    const yBase = HDR_H + Math.max(nodeInputs(n).length, nodeOutputs(n).length) * ROW_H + 4;
    intents.forEach((it, i) => {
      g.append("text").attr("x", 10).attr("y", yBase + 10 + i * 14)
        .attr("font-size", 9).attr("fill", it.description ? "#475569" : "#b6bcc4")
        .text(`${i + 1}. ${it.name || "…"}${it.description ? " — " + String(it.description).slice(0, 18) : "（无描述）"}`);
    });
  },
});
function ipGet(n, k) {
  const lp = n.data.inputs?.inputParameters || [];
  return lp.find(x => x.name === k)?.input?.value?.content ?? "";
}
function ipSet(n, k, v) {
  let lp = n.data.inputs.inputParameters || (n.data.inputs.inputParameters = []);
  let p = lp.find(x => x.name === k);
  if (!p) { p = { name: k, input: { type: "string", value: { type: "literal", content: "" } } }; lp.push(p); }
  p.input.value.content = v;
}
// 意图编辑器全局（注入脚本=全局作用域；name/description 双列）
function NODEP_INANO_set(i, k, v) { const n = findN(selNode); n.data.inputs.intents[i][k] = v; renderAll(); }
function NODEP_INANO_add()   { const n = findN(selNode); (n.data.inputs.intents = n.data.inputs.intents || []).push({ name: '新意图', description: '' }); showProps(); renderAll(); }
function NODEP_INANO_del(i)  { const n = findN(selNode); n.data.inputs.intents.splice(i, 1); showProps(); renderAll(); }

File: src/assets/nodes_builtin/intent_nano.py
"""NanoJev 快速意图路由节点插件（type intent_nano）：判别式意图分类——内置 Intent（type 22）的加强版。

与内置 intent（LLM 生成式：提示词→回复编号→解析）的区别：
· 判别式：NanoJev 0.6B 决策模型（Qwen3-0.6B backbone + 决策头）一次前向直接输出候选概率分布
  ——不生成文本，无编号解析歧义，CPU fp32 单次前向 ~1s、零 token 费用
· 可解释：输出完整概率分布 + 置信度；top1 置信度 < threshold 时视为不确定 → 走 default
  分支（不确定样本不硬路由——生成式 intent 做不到的软拒识）
· 意图描述直供模型：XML 里 <intent name="提问">用户想了解…</intent> 的描述体作为
  criteria 语义描述（内置 intent 只用 name 拼提示词）

服务链路：nanojev_server（http-task 型，127.0.0.1:8766；lfm_services.json 注册、
lfm_proxy --managed 托管）。插件直连 8766（省一层转发）；不可达时自动拉起
（detached + 等 health 最长 60s）；仍失败 → intent="" 走 default（不炸工作流）。

下游兼容：出口端口与内置 intent 同构（branch_N 按意图序，未命中 default）——
已有工作流可直接替换节点迁移。
"""
import json as _json
import subprocess as _subprocess
import time as _time
import urllib.request as _urlreq

PORT = 8766
BASE = f"http://127.0.0.1:{PORT}"
LAUNCH_CMD = ["python", r"D:\models\NanoJev\tools\nanojev_server.py",
              "--checkpoint", r"D:\models\NanoJev", "--port", str(PORT)]

PARAMS = [
    {"key": "query", "type": "string", "required": True,
     "desc": "待分类文本（通常 ref 上游 user_message 等）"},
    {"key": "intents", "type": "list", "required": True,
     "desc": "意图列表：name=意图名（对应 branch_N 出口），description=语义描述（供模型判别；空则用 name）"},
    {"key": "temperature", "type": "number", "required": False, "default": 1.0,
     "desc": "NanoJev 采样温度（默认 1.0；想更确定性可调低，如 0.3）"},
    {"key": "threshold", "type": "number", "required": False, "default": 0.0,
     "desc": "置信度阈值（0-1）：top1 概率低于它 → 走 default（软拒识不确定样本）"},
]


def _http_json(url, payload=None, timeout=15):
    if payload is None:
        req = _urlreq.Request(url)
    else:
        req = _urlreq.Request(url, data=_json.dumps(payload, ensure_ascii=False).encode("utf-8"),
                              method="POST")
        req.add_header("Content-Type", "application/json")
    with _urlreq.urlopen(req, timeout=timeout) as r:
        return _json.loads(r.read().decode("utf-8"))


def _ensure_server(timeout=60):
    """nanojev_server 探活；不通则按 lfm_services.json 的 cmd 拉起并等就绪。"""
    try:
        _http_json(f"{BASE}/api/health", timeout=4)
        return True
    except Exception:
        pass
    try:
        _subprocess.Popen(LAUNCH_CMD, creationflags=0x00000008 | 0x00000200,
                          stdout=_subprocess.DEVNULL, stderr=_subprocess.DEVNULL,
                          stdin=_subprocess.DEVNULL)
    except Exception:
        return False
    deadline = _time.time() + timeout
    while _time.time() < deadline:
        _time.sleep(2)
        try:
            _http_json(f"{BASE}/api/health", timeout=4)
            return True
        except Exception:
            continue
    return False


def _decide(query, criteria, instructions, temperature):
    """submit → 轮询 result → 返回 answers 或 None。"""
    body = {"states": [{
        "id": "route",
        "state": str(query),
        "questions": {"intent": {
            "type": "choice",
            "instructions": instructions,
            "criteria": criteria,
        }}}],
        "temperature": float(temperature or 1.0)}
    sub = _http_json(f"{BASE}/api/submit", body, timeout=20)
    rid = sub.get("request_id")
    if not rid:
        return None
    for _ in range(60):
        _time.sleep(1.0)
        res = _http_json(f"{BASE}/api/result/{rid}", timeout=10)
        if res.get("status") in ("done", "failed", "error"):
            if res.get("status") != "done":
                return None
            try:
                return res["result"]["states"][0]["answers"]["intent"]
            except Exception:
                return None
    return None


def _handle_intent_nano(node: dict, ctx) -> dict:
    from workflow_node_api import resolve_input_params

    inputs = node.get("data", {}).get("inputs", {})
    params = resolve_input_params(inputs.get("inputParameters", []), ctx)
    intents = [(i.get("name", "") or "", (i.get("description") or "").strip())
               for i in inputs.get("intents", []) if (i.get("name") or "").strip()]
    query = params.get("query") or next((v for v in params.values() if v), "")
    try:
        temperature = float(params.get("temperature") or 1.0)
    except Exception:
        temperature = 1.0
    try:
        threshold = float(params.get("threshold") or 0.0)
    except Exception:
        threshold = 0.0

    if not intents or not str(query).strip():
        return {"outputs": {"intent": "", "confidence": 0.0, "probabilities": {}, "raw": "",
                            "error": "query/intents 为空"}, "port": "default"}
    if not _ensure_server():
        return {"outputs": {"intent": "", "confidence": 0.0, "probabilities": {},
                            "raw": "", "error": "nanojev_server 不可达（8766 拉起失败）"},
                "port": "default"}

    criteria = {name: (desc or name) for name, desc in intents}
    instructions = str(params.get("instructions") or "判断输入文本属于哪个意图类别")
    ans = _decide(query, criteria, instructions, temperature)
    if not ans:
        return {"outputs": {"intent": "", "confidence": 0.0, "probabilities": {}, "raw": "",
                            "error": "nanojev 决策失败/超时"}, "port": "default"}

    probs = ans.get("probabilities") or {}
    choice = str(ans.get("choice") or "")
    conf = float(probs.get(choice, 0.0) or 0.0)
    out = {"intent": choice if conf >= threshold else "",
           "confidence": round(conf, 4),
           "probabilities": {k: round(float(v), 4) for k, v in probs.items()},
           "raw": _json.dumps(ans, ensure_ascii=False)[:2000], "error": ""}
    if out["intent"]:
        idx = next((i for i, (name, _d) in enumerate(intents) if name == out["intent"]), None)
        if idx is not None:
            return {"outputs": out, "port": f"branch_{idx}"}
    return {"outputs": out, "port": "default"}


def agt_node():
    return {"type": "intent_nano", "label": "Intent·Nano", "handler": _handle_intent_nano, "catalog": _CATALOG}


# ===== 节点目录条目（list_workflow_nodes / query_workflow_node 动态聚合自插件声明）=====
_CATALOG = {"name": "意图识别·NanoJev（判别式加强版）",
            "desc": "本地 NanoJev 0.6B 决策模型做意图分类：一次前向输出候选概率分布（~1s、零 token 费用、"
                    "无编号解析歧义），输出 intent/confidence/probabilities，top1 低于阈值走 default（软拒识）。"
                    "出口与内置 Intent 同构（branch_N/default），可直接替换迁移。",
            "xml": "<!-- NanoJev 判别式意图路由：描述体直供模型判别 -->\n"
                   "<node id=\"160001\" type=\"intent_nano\" title=\"意图路由\">\n"
                   "  <in name=\"query\" ref=\"100001.user_message\"/>\n"
                   "  <in name=\"temperature\" type=\"number\">1.0</in>\n"
                   "  <in name=\"threshold\" type=\"number\">0.35</in>\n"
                   "  <intent name=\"code\">要求编写、修改、调试代码</intent>\n"
                   "  <intent name=\"search\">要求检索资料、查文档、搜索信息</intent>\n"
                   "  <intent name=\"chat\">闲聊、打招呼或常识问答</intent>\n"
                   "  <out name=\"intent\" type=\"string\"/>\n"
                   "  <out name=\"confidence\" type=\"number\"/>\n"
                   "  <out name=\"probabilities\" type=\"object\"/>\n"
                   "</node>\n"
                   "<!-- 出口：branch_0(code) branch_1(search) branch_2(chat) default(未命中/低置信) -->"}
